Add memory_metrics to canister_status. (#5240)
dfinity/portal: docs/references/ic-interface-spec.md

-   A SHA256 hash of the module installed on the canister. This is `null` if the canister is empty.

> -   The actual memory usage of the canister.

Reviewer: I would keep this line and maybe make it a bit clearer that this refers to the total memory usage and additionally, create a new section that describes the detailed breakdown of memory types. As long as we keep both fields in the canister status result, both should be listed for consistency.

> -   Detailed breakdown of memory usage into individual components (in bytes):

Reviewer: High level comment: the interface spec uses the term "system state" mostly to refer to the system state tree which is different than what we call "system state" in the current implementation of the Internet Computer. Therefore, I would not necessarily try to capture this in the below but rather stay on a higher level describing what type of memory the field represents.

Additionally, for concepts that are covered elsewhere in the spec, we do not need to provide further clarifications imo. E.g. there's no need to explain what is contained in the canister history as the relevant section covers it.

    * `heap_size`: Represents the total Wasm memory usage of the execution state, which includes the heap memory used by the canister's WebAssembly code.

>     * `stable_size`: Indicates the total stable memory usage of the execution state, which is the memory that persists across canister upgrades.

Reviewer: ```suggestion
    * `stable_size`: Represents the stable memory usage of the canister, which is the memory that persists across canister upgrades.
```

    * `global_size`: The total global memory usage of the execution state, encompassing all global variables used by the canister.

>     * `wasm_binary_size`: The total memory occupied by the Wasm binary itself within the execution state.

Reviewer: ```suggestion
    * `wasm_binary_size`: Represents the memory occupied by the Wasm binary that is currently installed on the canister.
```

>     * `custom_sections_size`: The memory used by custom sections within the execution state, which may include additional metadata or configuration data.

Reviewer: ```suggestion
    * `custom_sections_size`: Represents the memory used by custom sections defined by the canister, which may include additional metadata or configuration data.
```

>     * `canister_history_size`: The total memory allocated for storing the canister's history within the system state, which includes records of past actions and changes.

Reviewer: ```suggestion
    * `canister_history_size`: Represents the memory used for storing the canister's history.
```

>     * `wasm_chunk_store_size`: The memory used by the Wasm chunk store within the system state, which is used to store large Wasm modules in chunks.

Reviewer: ```suggestion
    * `wasm_chunk_store_size`: Represents the memory used by the Wasm chunk store of the canister.
```

>     * `snapshot_size`: The total memory consumed by snapshots within the system state, which are used to capture the state of the canister at specific points in time.

Reviewer: ```suggestion
    * `snapshot_size`: Represents the memory consumed by all snapshots that belong to this canister.
```

-   A SHA256 hash of the module installed on the canister. This is `null` if the canister is empty.

-   The actual memory usage of the canister, representing the total memory consumed by the canister.

> -   A record containing detailed breakdown of memory usage into individual components.

Reviewer 2: ```suggestion
-   A record containing detailed breakdown of memory usage into individual components (see [Memory Metrics](#ic-canister_status-memory_metrics) for more details).
```

Only the controllers of the canister or the canister itself can request its status.

#### Memory Metrics {#ic-canister_status-memory_metrics}

    * `heap_size`: Represents the Wasm memory usage of the canister, i.e. the heap memory used by the canister's WebAssembly code.

>     * `stable_size`: Represents the stable memory usage of the canister, which is the memory that persists across canister upgrades.

Reviewer 2: wasm heap (or main) memory is also persisted for Motoko EOP

Reviewer: This is true. However, it would feel kinda weird to mention some special case for Motoko OP in the spec. What would you suggest? Maybe just drop the part "which is the memory that persists across canister upgrades" for stable memory? I guess we can argue that stable memory is sufficiently described in other sections of the spec and therefore here we do not need to overdo it.

Reviewer 2: Indeed, I'd drop `which is the memory that persists across canister upgrades.`

    * `global_size`: Represents the memory usage of the global variables that the canister is using.

>     * `wasm_binary_size`: Represents the memory occupied by the Wasm binary that is currently installed on the canister.

Reviewer 2: Let's point out that this is the size of the binary uploaded via `install_code` or `install_chunked_code`, e.g., the compressed size if the uploaded binary is gzipped.
